pkg/agent/multicast/mcast_route.go

```diff
@@ -0,0 +1,336 @@
+// by making these interfaces accept multicast traffic with multicast ip:mgroup.
+// https://tldp.org/HOWTO/Multicast-HOWTO-6.html#ss6.4
+func (c *MRouteClient) multicastInterfacesJoinMgroup(mgroup net.IP) error {
+	for _, config := range c.multicastInterfaceConfigs {
+		addrIP := config.IPv4Addr.IP.To4()
+		groupIP := mgroup.To4()
+		err := c.socket.MulticastInterfaceJoinMgroup(groupIP, addrIP, config.Name)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (c *MRouteClient) multicastInterfacesLeaveMgroup(mgroup net.IP) error {
+	for _, config := range c.multicastInterfaceConfigs {
+		addrIP := config.IPv4Addr.IP.To4()
+		groupIP := mgroup.To4()
+		err := c.socket.MulticastInterfaceLeaveMgroup(groupIP, addrIP, config.Name)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// processIGMPNocacheMsg reads igmpMsg from the multicast socket and configures
+// multicast route based on VIF value in the message.
+func (c *MRouteClient) processIGMPNocacheMsg(igmpMsg []byte) {
+	klog.V(2).InfoS("Received igmpMsg", "igmpMsg", igmpMsg)
+	msg, err := c.parseIGMPMsg(igmpMsg)
+	if err != nil {
+		klog.ErrorS(err, "Error parsing IGMP message")
+		return
+	}
+	if msg.VIF != c.internalInterfaceVIF {
+		// Skip inbound multicast traffic when there is no multicast receiver Pod
+		// listening the msg.Dst group.
+		status, ok, _ := c.groupCache.GetByKey(msg.Dst.String())
+		if !ok {
+			return
+		}
+		groupStatus := status.(*GroupMemberStatus)
+		if len(groupStatus.localMembers) == 0 {
+			return
+		}
+		// Prevent adding route entries for unrecognized VIF.
+		if len(c.externalInterfaceVIFs) < int(msg.VIF) {
+			klog.ErrorS(fmt.Errorf("error finding VIF"), "Adding inbound multicast route entry failed", "VIF", msg.VIF)
+			return
+		}
+		err := c.addInboundMrouteEntry(msg.Src, msg.Dst, msg.VIF)
+		if err != nil {
+			klog.ErrorS(err, "Adding inbound multicast route entry failed")
+		}
+	} else {
+		err = c.addOutboundMrouteEntry(msg.Src, msg.Dst)
+		if err != nil {
+			klog.ErrorS(err, "Adding outbound multicast route entry failed")
+		}
+	}
+}
+
+func (c *MRouteClient) deleteInboundMrouteEntryByGroup(group net.IP) (err error) {
+	klog.V(2).InfoS("Deleting multicast group", "group", group)
+	mEntries, _ := c.inboundRouteCache.ByIndex(GroupNameIndexName, group.String())
+	for _, route := range mEntries {
+		entry := route.(*inboundMulticastRouteEntry)
+		err := c.socket.DelMrouteEntry(net.ParseIP(entry.src), net.ParseIP(entry.group), entry.vif)
+		if err != nil {
+			return err
+		}
+		c.inboundRouteCache.Delete(route)
+	}
+	return nil
+}
+
+// addOutboundMrouteEntry configures multicast route from Antrea gateway to all the multicast interfaces,
+// allowing multicast sender Pods to send multicast traffic to external.
+func (c *MRouteClient) addOutboundMrouteEntry(src net.IP, group net.IP) (err error) {
+	klog.V(2).InfoS("Adding outbound multicast route entry", "src", src, "group", group, "vifs for oifs", c.externalInterfaceVIFs)
```

Comment:
don't include spaces in the key of structured log, do a test and you will see how it looks like.

Reply:
updated